File: scratch/review-observation-proof.py
"""Prepare disposable native output for the desktop/CLI review probe."""
from datetime import UTC, datetime
import json
from pathlib import Path
import tempfile

root = Path(tempfile.mkdtemp(prefix='loo293-observation-review-'))
for number, provider in enumerate(['claude', 'codex'], 1):
    run = f'run_{number:032x}'
    directory = root / 'runs' / run[4:6] / run
    directory.mkdir(parents=True)
    path = directory / 'native.jsonl'
    manifest = dict(schema_version=1, run_id=run, parent_run_id=None,
                    created_at=datetime.now(UTC).isoformat(), harness=provider, model=None,
                    surface='tui', cwd='/tmp', repo=None, worktree=None, skill=None,
                    subjects=[dict(selector='task:LOO-293', source='declared')],
                    launch=None, context=None, runtime_path=None, runtime_digest=None,
                    host='review', boot_id=None)
    receipt = dict(schema_version=1, provider_session_id='session', account_id=None,
                   native_source=dict(kind='jsonl', path=str(path)))
    (directory / 'manifest.json').write_text(json.dumps(manifest))
    (directory / 'provider-session.json').write_text(json.dumps(receipt))
    if provider == 'claude':
        initial = [dict(type='assistant', sessionId='session', uuid='call', message=dict(id='assistant', content=[
            dict(type='tool_use', id='call', name='bash', input=dict(command='echo review-proof'))]))]
        arrival = dict(type='user', sessionId='session', uuid='result', message=dict(content=[
            dict(type='tool_result', tool_use_id='call', content='review-proof')]))
    else:
        initial = [dict(type='session_meta', payload=dict(id='session')),
                   dict(type='response_item', payload=dict(id='call', type='function_call', call_id='call',
                        name='bash', arguments='echo review-proof'))]
        arrival = dict(type='response_item', payload=dict(id='result', type='function_call_output',
                       call_id='call', output='review-proof'))
    path.write_text(''.join(json.dumps(row) + '\n' for row in initial))
    (root / f'{provider}-arrival.jsonl').write_text(json.dumps(arrival) + '\n')
for number, provider in enumerate(['claude', 'codex'], 1):
    for ordinal in [1, 2]:
        text = f'{provider} arrival {ordinal}'
        if provider == 'claude':
            row = dict(type='assistant', sessionId='session', uuid=f'prose-{ordinal}',
                       message=dict(id=f'message-{ordinal}', content=[dict(type='text', text=text)]))
        else:
            row = dict(type='response_item', payload=dict(id=f'prose-{ordinal}', type='message',
                       role='assistant', content=[dict(type='output_text', text=text)]))
        (root / f'{provider}-prose-{ordinal}.jsonl').write_text(json.dumps(row) + '\n')
Path('/tmp/loo293-observation-proof-root').write_text(str(root))
print(root)

File: scratch/review-observation-proof.swift
#if os(macOS)
import AppKit
import Foundation
import SwiftUI
import Testing
import ViewInspector
@testable import Loopflow
@testable import LoopflowMac

@MainActor
@Suite("Watch observation CLI review", .serialized)
struct WatchObservationCLIReview {
    private let checkout = "/Users/jack/src/loopflow.restore-task-watching-with-live"
    private func query(home: String) -> RegistryQuery {
        RegistryQuery { args, _ in
            // Only the branch's early read-only Task dispatch is exercised.
            #expect(args.prefix(2) == ["task", "watch"] || args.prefix(2) == ["task", "output"])
            let process = Process()
            process.executableURL = URL(fileURLWithPath: checkout + "/target/debug/lf")
            process.arguments = args
            process.currentDirectoryURL = URL(fileURLWithPath: "/tmp")
            var env = ProcessInfo.processInfo.environment
            env["LF_CONTROL_HOME"] = home
            env["LF_CONTROL_DB_PATH"] = "/Users/jack/.lf/loopflow.db"
            env.removeValue(forKey: "LF_RUN_ID")
            process.environment = env
            let path = URL(fileURLWithPath: "/tmp/loo293-observation-query-\(UUID().uuidString)")
            FileManager.default.createFile(atPath: path.path, contents: nil)
            let output = try FileHandle(forWritingTo: path)
            defer { try? output.close(); try? FileManager.default.removeItem(at: path) }
            process.standardOutput = output
            process.standardError = FileHandle.nullDevice
            try process.run()
            process.waitUntilExit()
            #expect(process.terminationStatus == 0)
            return try String(contentsOf: path, encoding: .utf8)
        }
    }

    @Test func nativeArrivals() async throws {
        let home = try String(contentsOfFile: "/tmp/loo293-observation-proof-root", encoding: .utf8)
        let query = query(home: home)
        let store = TaskWatchStore()
        await store.refresh(issue: "LOO-293", query: query)
        await store.readOutput(issue: "LOO-293", query: query)
        #expect(store.outputError == nil)
        #expect(store.output.count == 2)
        #expect(store.output.allSatisfy { $0.rows.first?.title == "bash · running" })
        func append(_ provider: String, _ suffix: String) throws {
            let run = String(format: "run_%032x", provider == "claude" ? 1 : 2)
            let file = try FileHandle(forWritingTo: URL(fileURLWithPath: home + "/runs/00/" + run + "/native.jsonl"))
            defer { try? file.close() }
            try file.seekToEnd()
            try file.write(contentsOf: Data(contentsOf: URL(fileURLWithPath: home + "/\(provider)-\(suffix).jsonl")))
        }
        for (provider, suffix) in [("claude", "prose-1"), ("codex", "prose-1"), ("claude", "prose-2")] {
            try append(provider, suffix)
            await store.readOutput(issue: "LOO-293", query: query)
            #expect(store.outputError == nil)
        }
        let arrivals = store.outputGroups.filter { !$0.history }
        #expect(arrivals.map { $0.source.provider } == ["claude", "codex", "claude"])
        #expect(arrivals.flatMap { $0.rows.map(\.text) } == ["claude arrival 1", "codex arrival 1", "claude arrival 2"])
        let blocks = store.outputGroups.map(\.id)
        let before = store.latestOutputRow
        try append("claude", "arrival")
        await store.readOutput(issue: "LOO-293", query: query)
        #expect(store.outputError == nil)
        #expect(store.outputGroups.map(\.id) == blocks)
        let tool = try #require(store.output.first { $0.source.provider == "claude" }?.rows.first)
        #expect(tool.title == "bash · completed")
        #expect(tool.text.contains("echo review-proof") && tool.text.hasSuffix("\nreview-proof"))
        #expect(store.latestOutputRow?.row == tool.id)
        #expect(store.latestOutputRow != before)
        let target = store.latestOutputRow
        let rows = store.output.map { $0.rows.map(\.text) }
        await store.readOutput(issue: "LOO-293", query: query)
        #expect(store.output.map { $0.rows.map(\.text) } == rows)
        #expect(store.latestOutputRow == target)
        #expect(store.outputGroups.map(\.id) == blocks)
        store.inspectRun(store.output[0].source.runId)
        #expect(store.visibleOutput.count == 1)
        let view = TaskWatchOutputView(store: store, onHistory: {}, onFollow: { store.followLive() }, onReload: {})
        try view.inspect().find(button: "Follow live").tap()
        #expect(store.visibleOutput.count == 2)
        try render(TaskWatchView(issue: "LOO-293", store: store))
        print("PASS real CLI -> RegistryQuery -> Store -> View: A/B/A arrivals, original tool row updated, exact Follow target, quiet deduplication, Run filter and Follow action")
    }

    @Test func configuredHistory() async throws {
        let store = TaskWatchStore()
        let query = query(home: "/Users/jack/.lf")
        await store.refresh(issue: "LOO-293", query: query)
        await store.readOutput(issue: "LOO-293", query: query)
        #expect(store.error == nil && store.outputError == nil)
        #expect(store.snapshot?.taskId == "task_97e04da37e5e4231bde30f4d1d93a5fa")
        #expect(store.output.contains { !$0.rows.isEmpty })
        print("PASS configured passive history: sources=\(store.output.count), rows=\(store.output.reduce(0) { $0 + $1.rows.count }), invocations=\(store.snapshot?.invocations.count ?? 0)")
    }

    private func render(_ view: TaskWatchView) throws {
        let host = NSHostingView(rootView: view.frame(width: 1100, height: 850).environment(\.palette, LoopflowPalette.light))
        let window = NSWindow(contentRect: NSRect(x: 0, y: 0, width: 1100, height: 850), styleMask: [.borderless], backing: .buffered, defer: false)
        window.isReleasedWhenClosed = false
        window.appearance = NSAppearance(named: .aqua)
        window.contentView = host
        defer { window.close() }
        host.layoutSubtreeIfNeeded()
        window.displayIfNeeded()
        let bitmap = try #require(host.bitmapImageRepForCachingDisplay(in: host.bounds))
        host.cacheDisplay(in: host.bounds, to: bitmap)
        try #require(bitmap.representation(using: .png, properties: [:])).write(to: URL(fileURLWithPath: "/tmp/loo293-observation-cli-view.png"))
    }
}
#endif
